File: src/ios/OIDCPlugin.swift
import Foundation
import Cordova
import AppAuth

@objc(OIDCPlugin)
class OIDCPlugin: CDVPlugin {
    
    let authService = AuthService.shared

	func validSession(command: CDVInvokedUrlCommand) {

		AuthService.shared.validSession { result in

			var pluginResult: CDVPluginResult


			switch result {

			case .valid:

				pluginResult = CDVPluginResult(
					status: CDVCommandStatus_OK,
					messageAs: true
				)


			case .invalid:

				pluginResult = CDVPluginResult(
					status: CDVCommandStatus_OK,
					messageAs: false
				)


			case .unavailable:

				pluginResult = CDVPluginResult(
					status: CDVCommandStatus_ERROR,
					messageAs: "Network unavailable"
				)
			}


			self.commandDelegate.send(
				pluginResult,
				callbackId: command.callbackId
			)
		}
	}
    
    @objc(login:)
    func login(command: CDVInvokedUrlCommand) {
        guard let config = command.arguments.first as? [String: Any],
              let issuerStr = config["issuer"] as? String,
              let issuer = URL(string: issuerStr),
              let clientId = config["clientId"] as? String,
              let scopesStr = config["scopes"] as? String,
			  let prompt = config["prompt"] as? Bool,
			  let slo = config["slo"] as? Bool,
              let redirectURIStr = config["redirectURI"] as? String,
              let redirectURI = URL(string: redirectURIStr) else {
            
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Ungültige Konfigurationsparameter")
            self.commandDelegate.send(result, callbackId: command.callbackId)
            return
        }
        
        let scopes = scopesStr.components(separatedBy: " ")
        
        // ViewController von Cordova nutzen
        authService.login(presenting: self.viewController, issuer: issuer, clientId: clientId, scopes: scopes, redirectURI: redirectURI, prompt: prompt, slo: slo) { success in
            let result = CDVPluginResult(status: success ? CDVCommandStatus_OK : CDVCommandStatus_ERROR, messageAs: success ? "Login erfolgreich" : "Login fehlgeschlagen")
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(logout:)
    func logout(command: CDVInvokedUrlCommand) {
        guard let config = command.arguments.first as? [String: Any],
              let redirectURIStr = config["redirectURI"] as? String,
              let redirectURI = URL(string: redirectURIStr) else {
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "redirectURI fehlt in Konfiguration")
            self.commandDelegate.send(result, callbackId: command.callbackId)
            return
        }
        
        // idToken holen für ADFS Logout
        let idToken = authService.getIdToken()
        
        authService.logout(presenting: self.viewController, idToken: idToken, postLogoutRedirectURI: redirectURI) { success in
          let result = CDVPluginResult(status: success ? CDVCommandStatus_OK : CDVCommandStatus_ERROR, messageAs: success ? "Logout erfolgreich" : "Logout fehlgeschlagen")
          self.commandDelegate.send(result, callbackId: command.callbackId)
		}
    }
    
    @objc(hasAccount:)
    func hasAccount(command: CDVInvokedUrlCommand) {
        
	if !authService.isSessionValid() {

		// Session ist global invalid
		let result = CDVPluginResult(
			status: CDVCommandStatus_ERROR,
			messageAs: "Benutzer ist abgemeldet."
		)

		self.commandDelegate.send(result, callbackId: command.callbackId)
		return
	}

	let isAuthenticated = authService.isAuthenticated

        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: isAuthenticated)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    @objc(getToken:)
    func getToken(command: CDVInvokedUrlCommand) {
        authService.performAuthenticatedRequest { accessToken, idToken, error in
            if let error = error {
                let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: error.localizedDescription)
                self.commandDelegate.send(result, callbackId: command.callbackId)
                return
            }
            
            var tokens: [String: String] = [:]
            if let accToken = accessToken { tokens["access_token"] = accToken }
            if let idTkn = idToken { tokens["id_token"] = idTkn }
            
            let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: tokens)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
	@objc(usersinfos:)
	func usersinfos(command: CDVInvokedUrlCommand) {

		guard let config = command.arguments.first as? [String: Any] else {
			let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Ungültige Konfigurationsparameter")
			self.commandDelegate.send(result, callbackId: command.callbackId)
			return
		}
		
		// Parameter auslesen (LDAP oder Fallback OIDC)
		let ldapInfoUrlStr = config["ldapinfourl"] as? String
		let userClaimKey = config["userclaim"] as? String ?? "sub" // Fallback auf "sub", falls nicht angegeben
		let issuerStr = config["issuer"] as? String

		// 1. Token intern holen
		authService.performAuthenticatedRequest { [weak self] accessToken, idToken, error in
			guard let self = self else { return }

			if let error = error {
				self.sendCordovaError(error.localizedDescription, callbackId: command.callbackId)
				return
			}

			guard let accessToken = accessToken else {
				self.sendCordovaError("Kein Access Token vorhanden", callbackId: command.callbackId)
				return
			}

			// --- NEUER LDAP FLOW ---
			if let ldapStr = ldapInfoUrlStr, !ldapStr.isEmpty {
				
				// URL zusammenbauen wie im Java Code
				var finalLdapUrlStr = ldapStr
				if !finalLdapUrlStr.hasSuffix("ldap/lookup") {
					finalLdapUrlStr += finalLdapUrlStr.hasSuffix("/") ? "ldap/lookup" : "/ldap/lookup"
				}
				
				guard let ldapUrl = URL(string: finalLdapUrlStr) else {
					self.sendCordovaError("Ungültige LDAP URL", callbackId: command.callbackId)
					return
				}
				
				guard let idToken = idToken else {
					self.sendCordovaError("Für den LDAP Request wird ein ID Token benötigt", callbackId: command.callbackId)
					return
				}
				
				// Claim aus dem ID Token extrahieren
				let filterValue = self.extractClaim(from: idToken, claimKey: "persnr") ?? ""
				
				// Request mit Retry-Logik starten
				self.performLDAPRequest(url: ldapUrl, accessToken: accessToken, filter: filterValue, retriesLeft: 3) { result in
					switch result {
					case .success(let parsedData):
						if let jsonData = try? JSONSerialization.data(withJSONObject: parsedData),
						let jsonString = String(data: jsonData, encoding: .utf8) {
							let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: jsonString)
							self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
						} else {
							self.sendCordovaError("Fehler beim Serialisieren der LDAP Daten", callbackId: command.callbackId)
						}
					case .failure(let error):
						self.sendCordovaError(error.localizedDescription, callbackId: command.callbackId)
					}
				}
				
			} 
			// --- ALTER OIDC FLOW (Fallback) ---
			else if let issuerStr = issuerStr, let issuer = URL(string: issuerStr) {
				
				let userInfoURL = issuer.appendingPathComponent("userinfo")
				var request = URLRequest(url: userInfoURL)
				request.httpMethod = "GET"
				request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
				request.setValue("application/json", forHTTPHeaderField: "Accept")

				let task = URLSession.shared.dataTask(with: request) { data, response, error in
					if let error = error {
						self.sendCordovaError(error.localizedDescription, callbackId: command.callbackId)
						return
					}

					guard let httpResponse = response as? HTTPURLResponse, let data = data else {
						self.sendCordovaError("Keine gültige Antwort", callbackId: command.callbackId)
						return
					}

					if !(200...299).contains(httpResponse.statusCode) {
						let body = String(data: data, encoding: .utf8) ?? ""
						self.sendCordovaError("HTTP \(httpResponse.statusCode): \(body)", callbackId: command.callbackId)
						return
					}

					let jsonString = String(data: data, encoding: .utf8) ?? "{}"
					let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: jsonString)
					self.commandDelegate.send(result, callbackId: command.callbackId)
				}
				task.resume()
				
			} else {
				self.sendCordovaError("Weder LDAP-URL noch Issuer-URL gefunden", callbackId: command.callbackId)
			}
		}
	}

	// MARK: - Helper Methods

	/// Hilfsfunktion um Cordova Fehler zu senden und Boilerplate zu reduzieren
	private func sendCordovaError(_ message: String, callbackId: String) {
		let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: message)
		self.commandDelegate.send(result, callbackId: callbackId)
	}

	/// Führt den LDAP POST-Request mit Retry-Logik (3 Versuche, 3 Sekunden Pause) aus
	private func performLDAPRequest(url: URL, accessToken: String, filter: String, retriesLeft: Int, completion: @escaping (Result<[String: Any], Error>) -> Void) {
		
		let attributes = [
			"samaccountname", "company", "mail", "facsimileTelephoneNumber", "telephoneNumber",
			"mobile", "streetAddress", "postalCode", "l", "displayName", "sn", "givenName",
			"department", "extensionAttribute2", "extensionattribute1", "extensionAttribute4",
			"jpegPhoto", "thumbnailPhoto", "office"
		]
		
		let jsonParams: [String: Any] = [
			"filter": filter,
			"attributes": attributes
		]
		
		var request = URLRequest(url: url)
		request.httpMethod = "POST"
		request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
		request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
		request.setValue("application/json", forHTTPHeaderField: "Accept")
		request.httpBody = try? JSONSerialization.data(withJSONObject: jsonParams)
		
		let task = URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
			guard let self = self else { return }
			
			let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
			let isSuccess = (200...299).contains(statusCode)
			
			if isSuccess, let data = data {
				if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
				let parsedData = self.parseInfosFromLdap(json) {
					completion(.success(parsedData))
					return
				}
			}
			
			// Retry Logic
			if retriesLeft > 1 {
				DispatchQueue.global().asyncAfter(deadline: .now() + 3.0) {
					self.performLDAPRequest(url: url, accessToken: accessToken, filter: filter, retriesLeft: retriesLeft - 1, completion: completion)
				}
			} else {
				let errorMsg = error?.localizedDescription ?? "Fehler beim Abrufen der UserInfo-Daten von LDAP"
				completion(.failure(NSError(domain: "LDAPService", code: -1, userInfo: [NSLocalizedDescriptionKey: errorMsg])))
			}
		}
		task.resume()
	}

	/// Parst die LDAP Antwort in ein flaches Dictionary
	private func parseInfosFromLdap(_ response: [String: Any]) -> [String: Any]? {
		guard let results = response["results"] as? [[String: Any]],
			let firstResult = results.first,
			let attributes = firstResult["attributes"] as? [[String: Any]] else {
			return nil
		}
		
		var resultDict: [String: Any] = [:]
		
		if let userId = firstResult["id"] {
			resultDict["user"] = userId
		}
		
		for attribute in attributes {
			if let name = attribute["name"] as? String, let value = attribute["value"] {
				resultDict[name.lowercased()] = value
			}
		}
		
		return resultDict
	}

	/// Extrahiert einen bestimmten Claim (z.B. Mail, SAMAccountName) aus dem JWT (ID Token)
	private func extractClaim(from jwt: String, claimKey: String) -> String? {
		let segments = jwt.components(separatedBy: ".")
		guard segments.count > 1 else { return nil }
		
		// Base64Url Decoding fixen (Padding auffüllen)
		var base64String = segments[1]
			.replacingOccurrences(of: "-", with: "+")
			.replacingOccurrences(of: "_", with: "/")
		
		let length = Double(base64String.lengthOfBytes(using: .utf8))
		let requiredLength = 4 * ceil(length / 4.0)
		let paddingLength = requiredLength - length
		if paddingLength > 0 {
			let padding = String(repeating: "=", count: Int(paddingLength))
			base64String += padding
		}
		
		guard let payloadData = Data(base64Encoded: base64String, options: .ignoreUnknownCharacters),
			let payload = try? JSONSerialization.jsonObject(with: payloadData, options: []) as? [String: Any] else {
			return nil
		}
		
		return payload[claimKey] as? String
	}
	
    // SEHR WICHTIG: Fängt die Callback-URL ab, wenn die AppAuth-Webview zurück zur App leitet
    override func handleOpenURL(_ notification: Notification) {
		print("AUTHCALLBACK")
        guard let url = notification.object as? URL else { return }
        _ = authService.resumeExternalUserAgentFlow(with: url)
    }
}

File: src/android/OIDCPlugin.java
package de.mopsdom.oidc.cordova;

import android.content.Intent;
import android.net.Uri;
import android.util.Base64;
import org.apache.cordova.*;
import org.json.*;
import net.openid.appauth.*;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.*;

public class OIDCPlugin extends CordovaPlugin {
    private static final int RC_AUTH = 100;
    private static final int RC_END_SESSION = 101;

    private AuthService authService;
    private CallbackContext callbackContext;
    private AuthorizationService appAuthService;

    @Override
    protected void pluginInitialize() {
        this.authService = AuthService.getInstance(cordova.getActivity());
        this.appAuthService = new AuthorizationService(cordova.getActivity());
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {

        JSONObject config = args.optJSONObject(0);
        if ("login".equals(action)) {
            login(config,callbackContext);
            return true;
        } else if ("logout".equals(action)) {
            this.callbackContext = callbackContext;
            logout(config);
            return true;
        } else if ("isAuthenticated".equals(action)) {
            isAuthenticated(callbackContext);
            return true;
        } else if ("getToken".equals(action)) {
            getToken(callbackContext);
            return true;
        } else if ("usersinfos".equals(action)) {
            usersinfos(config,callbackContext);
            return true;
        }
        return false;
    }

    private void login(JSONObject config, CallbackContext currentCallbackContext) throws JSONException {
        String issuerStr = config.getString("issuer");
        String clientId = config.getString("clientId");
        String scopesStr = config.getString("scopes");
        boolean prompt = config.optBoolean("prompt", false);
        boolean slo = config.optBoolean("slo", false);
        String redirectURIStr = config.getString("redirectURI");

        Uri issuerUri = Uri.parse(issuerStr);
        Uri redirectUri = Uri.parse(redirectURIStr);
        List<String> scopes = Arrays.asList(scopesStr.split(" "));

        PluginResult pluginResult = new PluginResult(PluginResult.Status.NO_RESULT);
        pluginResult.setKeepCallback(true);
        currentCallbackContext.sendPluginResult(pluginResult);

        this.callbackContext = currentCallbackContext;

        AuthorizationServiceConfiguration.fetchFromIssuer(issuerUri, (serviceConfiguration, ex) -> {
            if (ex != null) {
                android.util.Log.e("OIDC_PROXY", "Discovery Fehler Details: ", ex);
                callbackContext.error("Discovery failed: " + ex.getLocalizedMessage());
                return;
            }

            boolean forcePrompt = slo ? !authService.isSessionValid() || prompt : prompt;
            AuthorizationRequest.Builder requestBuilder = new AuthorizationRequest.Builder(
                    serviceConfiguration, clientId, ResponseTypeValues.CODE, redirectUri)
                    .setScopes(scopes);

            if (forcePrompt) {
                requestBuilder.setPrompt("login");
            }

            AuthorizationRequest request = requestBuilder.build();
            Intent authIntent = appAuthService.getAuthorizationRequestIntent(request);

            // Startet den Custom-Tab/Browser Flow über Cordova
            cordova.startActivityForResult(this, authIntent, RC_AUTH);
        });
    }

    private void logout(JSONObject config) throws JSONException {
        String redirectURIStr = config.getString("redirectURI");
        Uri redirectUri = Uri.parse(redirectURIStr);

        if (authService.getAuthState() == null || authService.getAuthState().getAuthorizationServiceConfiguration() == null) {
            authService.clearState();
            authService.globalStore.setLogout(new Date().getTime());
            callbackContext.success("Logout erfolgreich (Lokal)");
            return;
        }

        AuthorizationServiceConfiguration discovery = authService.getAuthState().getAuthorizationServiceConfiguration();

        // Android AppAuth benötigt oft die manuelle EndSession-Konfiguration, falls nicht im Metadaten-Objekt automatisch gemappt
        if (discovery.endSessionEndpoint == null) {
            authService.clearState();
            authService.globalStore.setLogout(new Date().getTime());
            callbackContext.success("Logout erfolgreich");
            return;
        }

        String idToken = authService.getIdToken();
        EndSessionRequest endSessionRequest = new EndSessionRequest.Builder(discovery)
                .setIdTokenHint(idToken)
                .setPostLogoutRedirectUri(redirectUri)
                .build();

        Intent endSessionIntent = appAuthService.getEndSessionRequestIntent(endSessionRequest);
        cordova.startActivityForResult(this, endSessionIntent, RC_END_SESSION);
    }

    private void isAuthenticated(CallbackContext callbackContext) {
        if (!authService.isSessionValid()) {
            callbackContext.error("Benutzer ist abgemeldet.");
            return;
        }
        callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, authService.isAuthenticated()));
    }

    private void getToken(CallbackContext callbackContext) {
        authService.performAuthenticatedRequest((accessToken, idToken, error) -> {
            if (error != null) {
                callbackContext.error(error.getLocalizedMessage());
                return;
            }
            try {
                JSONObject tokens = new JSONObject();
                if (accessToken != null) tokens.put("access_token", accessToken);
                if (idToken != null) tokens.put("id_token", idToken);
                callbackContext.success(tokens);
            } catch (JSONException e) {
                callbackContext.error(e.getLocalizedMessage());
            }
        });
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == RC_AUTH) {
            AuthorizationResponse response = AuthorizationResponse.fromIntent(data);
            AuthorizationException ex = AuthorizationException.fromIntent(data);

            if (response != null) {
                // Token Exchange (Authorization Code grant) ausführen
                appAuthService.performTokenRequest(response.createTokenExchangeRequest(), (tokenResponse, exception) -> {
                    if (tokenResponse != null) {
                        authService.createNewAuthState(response, ex);
                        authService.updateAuthStateAfterTokenResponse(tokenResponse, exception);
                        callbackContext.success("Login erfolgreich");
                    } else {
                        callbackContext.error(exception != null ? exception.getLocalizedMessage() : "Token exchange failed");
                    }
                });
            } else {
                callbackContext.error(ex != null ? ex.getLocalizedMessage() : "Login abgebrochen");
            }
        } else if (requestCode == RC_END_SESSION) {
            // Android EndSession-Rückmeldung validieren
            authService.globalStore.setLogout(new Date().getTime());
            authService.clearState();
            callbackContext.success("Logout erfolgreich");
        }
    }

    // --- USERINFO & LDAP FLOW ---

    private void usersinfos(JSONObject config,CallbackContext callbackContext) throws JSONException {
        String ldapInfoUrlStr = config.optString("ldapinfourl", null);
        String userClaimKey = config.optString("userclaim", "sub");
        String issuerStr = config.optString("issuer", null);

        authService.performAuthenticatedRequest((accessToken, idToken, error) -> {
            if (error != null) {
                callbackContext.error(error.getLocalizedMessage());
                return;
            }
            if (accessToken == null) {
                callbackContext.error("Kein Access Token vorhanden");
                return;
            }

            if (ldapInfoUrlStr != null && !ldapInfoUrlStr.isEmpty()) {
                // LDAP Flow
                String finalLdapUrl = ldapInfoUrlStr;
                if (!finalLdapUrl.endsWith("ldap/lookup")) {
                    finalLdapUrl += finalLdapUrl.endsWith("/") ? "ldap/lookup" : "/ldap/lookup";
                }
                if (idToken == null) {
                    callbackContext.error("Für den LDAP Request wird ein ID Token benötigt");
                    return;
                }

                String filterValue = extractClaim(idToken, "persnr");
                performLDAPRequest(finalLdapUrl, accessToken, filterValue, 3,callbackContext);

            } else if (issuerStr != null) {
                // OIDC Fallback Flow
                performOIDCFallback(issuerStr, accessToken,callbackContext);
            } else {
                callbackContext.error("Weder LDAP-URL noch Issuer-URL gefunden");
            }
        });
    }

    private void performLDAPRequest(String urlStr, String accessToken, String filter, int retriesLeft,CallbackContext callbackContext) {
        cordova.getThreadPool().execute(() -> {
            try {
                URL url = new URL(urlStr);
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                conn.setRequestMethod("POST");
                conn.setRequestProperty("Authorization", "Bearer " + accessToken);
                conn.setRequestProperty("Content-Type", "application/json; charset=utf-8");
                conn.setRequestProperty("Accept", "application/json");
                conn.setDoOutput(true);

                JSONArray attributes = new JSONArray(Arrays.asList(
                    "samaccountname", "company", "mail", "facsimileTelephoneNumber", "telephoneNumber",
                    "mobile", "streetAddress", "postalCode", "l", "displayName", "sn", "givenName",
                    "department", "extensionAttribute2", "extensionattribute1", "extensionAttribute4",
                    "jpegPhoto", "thumbnailPhoto", "office"
                ));

                JSONObject jsonParams = new JSONObject();
                jsonParams.put("filter", filter);
                jsonParams.put("attributes", attributes);

                try (OutputStream os = conn.getOutputStream()) {
                    byte[] input = jsonParams.toString().getBytes(StandardCharsets.UTF_8);
                    os.write(input, 0, input.length);
                }

                int statusCode = conn.getResponseCode();
                if (statusCode >= 200 && statusCode < 300) {
                    BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
                    StringBuilder response = new StringBuilder();
                    String responseLine;
                    while ((responseLine = br.readLine()) != null) {
                        response.append(responseLine.trim());
                    }
                    JSONObject parsedData = parseInfosFromLdap(new JSONObject(response.toString()));
                    callbackContext.success(parsedData.toString());
                } else {
                    throw new IOException("HTTP Error " + statusCode);
                }
            } catch (Exception e) {
                if (retriesLeft > 1) {
                    try { Thread.sleep(3000); } catch (InterruptedException ignored) {}
                    performLDAPRequest(urlStr, accessToken, filter, retriesLeft - 1,callbackContext);
                } else {
                    callbackContext.error("LDAP Error: " + e.getLocalizedMessage());
                }
            }
        });
    }

    private void performOIDCFallback(String issuerStr, String accessToken,CallbackContext callbackContext) {
        cordova.getThreadPool().execute(() -> {
            try {
                URL url = new URL(issuerStr + (issuerStr.endsWith("/") ? "userinfo" : "/userinfo"));
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                conn.setRequestMethod("GET");
                conn.setRequestProperty("Authorization", "Bearer " + accessToken);
                conn.setRequestProperty("Accept", "application/json");

                int statusCode = conn.getResponseCode();
                InputStream is = (statusCode >= 200 && statusCode < 300) ? conn.getInputStream() : conn.getErrorStream();
                BufferedReader br = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8));
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = br.readLine()) != null) response.append(line.trim());

                if (statusCode >= 200 && statusCode < 300) {
                    callbackContext.success(response.toString());
                } else {
                    callbackContext.error("HTTP " + statusCode + ": " + response.toString());
                }
            } catch (Exception e) {
                callbackContext.error(e.getLocalizedMessage());
            }
        });
    }

    private JSONObject parseInfosFromLdap(JSONObject response) throws JSONException {
        JSONArray results = response.optJSONArray("results");
        if (results == null || results.length() == 0) return new JSONObject();

        JSONObject firstResult = results.getJSONObject(0);
        JSONArray attributes = firstResult.optJSONArray("attributes");
        JSONObject resultDict = new JSONObject();

        if (firstResult.has("id")) {
            resultDict.put("user", firstResult.get("id"));
        }

        if (attributes != null) {
            for (int i = 0; i < attributes.length(); i++) {
                JSONObject attr = attributes.getJSONObject(i);
                String name = attr.optString("name");
                Object value = attr.opt("value");
                if (name != null) {
                    resultDict.put(name.toLowerCase(), value);
                }
            }
        }
        return resultDict;
    }

    private String extractClaim(String jwt, String claimKey) {
        try {
            String[] parts = jwt.split("\\.");
            if (parts.length < 2) return "";
            byte[] decodedBytes = Base64.decode(parts[1], Base64.URL_SAFE);
            JSONObject payload = new JSONObject(new String(decodedBytes, StandardCharsets.UTF_8));
            return payload.optString(claimKey, "");
        } catch (Exception e) {
            return "";
        }
    }
}
